### PR TITLE
Add validation for project details import

### LINKED ISSUE #4632: Add validation for project details import
## Acceptance

- [x] Given I've generated a QR code with project details
And the project icon value is something longer than one character
When I try to add a project using the QR code
Then the project imports successfully
And the project icon is just the first character of the value in the QR code

- [x] Given I've generated a QR code with project details
And the project color value is not a valid color hex code
When I try to add a project using the QR code
Then the project imports successfully
And the project color is based on the name

- [x] Given I've generated a QR code with project details
And only the project name is included
When I try to add a project using the QR code
Then the project imports successfully
And the project icon is based on the name (not the server URL)
And the project color is based on the name (not the server URL)

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #4632 

#### What has been done to verify that this works as intended?
I tested this change manually and improved automated tests.

#### Why is this the best possible solution? Were any other approaches considered?
It has been implemented according to what we agreed in the issue.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
This will be tested later as a whole feature.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/projects/ProjectDetailsCreator.kt [hunk 1]
```diff
@@ -5,32 +5,74 @@ import androidx.core.content.ContextCompat
...
     }
+
+    private fun isProjectColorValid(hexColor: String): Boolean {
+        return Pattern
+            .compile("^#([a-fA-F0-9]{6}|[a-fA-F0-9]{3})\$", Pattern.CASE_INSENSITIVE)
+            .matcher(hexColor)
+            .matches()
+    }
+
+    private fun getFirstSing(value: String): String {
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/projects/ProjectDetailsCreator.kt]: I think the method name is a typo here?

File: collect_app/src/test/java/org/odk/collect/android/projects/ProjectDetailsCreatorTest.kt [hunk 2]
```diff
@@ -15,43 +15,87 @@ class ProjectDetailsCreatorTest {
...
 
     @Test
-    fun `Project icon should be generated from url`() {
-        assertThat(projectDetailsCreator.getProject("https://my-project.com").icon, `is`("M"))
-        assertThat(projectDetailsCreator.getProject("https://your-project.com/one").icon, `is`("Y"))
-        assertThat(projectDetailsCreator.getProject("http://www.my-project.com").icon, `is`("W"))
-        assertThat(projectDetailsCreator.getProject("").icon, `is`("P")) // default project icon for invalid urls
-        assertThat(projectDetailsCreator.getProject("something").icon, `is`("P")) // default project icon for invalid urls
+    fun `If project name is not included in project details or empty should be generated based on url`() {
+        // Case1: not included
```
Comment (seadowg) [inline collect_app/src/test/java/org/odk/collect/android/projects/ProjectDetailsCreatorTest.kt]: I think we should pull all these "cases" out in to separate tests

**Review by seadowg — CHANGES_REQUESTED** [review]
Just a couple of tweaks here, but all looks great. Was able to try it out by hacking the QR generation code.

File: collect_app/src/test/java/org/odk/collect/android/projects/ProjectDetailsCreatorTest.kt (hunk 2 above)
Comment (grzesiek2010) [reply collect_app/src/test/java/org/odk/collect/android/projects/ProjectDetailsCreatorTest.kt]: You mean never have two cases in one test or what?

Comment (seadowg) [reply collect_app/src/test/java/org/odk/collect/android/projects/ProjectDetailsCreatorTest.kt]: I just think if we feel like it's worth the comments, then we can just pull them out into separate tests and use the test names instead.

Comment (grzesiek2010) [reply collect_app/src/test/java/org/odk/collect/android/projects/ProjectDetailsCreatorTest.kt]: Done.

File: collect_app/src/main/java/org/odk/collect/android/projects/ProjectDetailsCreator.kt (hunk 1 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/projects/ProjectDetailsCreator.kt]: Done.

**Review by seadowg — APPROVED** [review]

**Pull request merged**